plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2819,33 +2831,37 @@ void producer_plugin_impl::switch_to_read_window() {
    app().executor().set_to_read_window();
    chain_plug->chain().set_db_read_only_mode();
    _received_block = false;
-
-   // start a read-only transaction execution task in each thread in the thread pool
-   _ro_num_active_trx_exec_tasks = _ro_thread_pool_size;
-   for (auto i = 0; i < _ro_thread_pool_size; ++i ) {
-      _ro_trx_exec_tasks_fut.emplace_back( post_async_task( _ro_thread_pool.get_executor(), [self = this] () {
-         return self->read_only_trx_execution_task();
-      }) );
+   {
+      std::lock_guard<std::mutex> g(_switch_window_mtx);
+      _switch_window_cond.notify_all();
    }
+   // read-only threads decide when to switch back to write window
+}
 
-   auto expire_time = boost::posix_time::microseconds(_ro_read_window_time_us.count());
-   _ro_read_window_timer.expires_from_now( expire_time );
-   _ro_read_window_timer.async_wait( app().executor().wrap(  // stay on app thread
-      priority::high,
-      exec_queue::read_only_trx_safe,
-      [weak_this = weak_from_this()]( const boost::system::error_code& ec ) {
-         auto self = weak_this.lock();
-         if( self && ec != boost::asio::error::operation_aborted ) {
-            // use future to make sure all read-only tasks finished before switching to write window
-            for ( auto& task: self->_ro_trx_exec_tasks_fut ) {
-               task.get();
-            }
-            self->_ro_trx_exec_tasks_fut.clear();
-            self->switch_to_write_window();
-          } else if ( self ) {
-             self->_ro_trx_exec_tasks_fut.clear();
-          }
-       }));
+// Called from a read only trx thread. Run in parallel with main and other read only trx threads
+void producer_plugin_impl::start_read_only_thread() {
+   chain::controller& chain = chain_plug->chain();
+   chain.init_thread_local_data();
+
+   while ( 1 ) {
```

Comment:
The `named_thread_pool` does not have the opportunity to call `_ioc.run()`. That works now because you don't post anything to the `_ro_thread_pool` executor, but seems rather odd.